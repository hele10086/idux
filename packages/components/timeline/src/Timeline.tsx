/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { TimelinePlacement, TimelineProps } from './types'
import type { VNode } from 'vue'

import { cloneVNode, defineComponent, h } from 'vue'

import { IxIcon } from '@idux/components/icon'

import IxTimelineItem from './TimelineItem'
import { timelineProps } from './types'

const timelinePrefixCls = 'ix-timeline'
const itemPrefixCls = 'ix-timeline-item'

export default defineComponent({
  name: 'IxTimeline',
  props: timelineProps,
  setup(props, { slots }) {
    return () => {
      const pendingSlots = slots?.pending?.()
      const pendingDotSlots = slots?.pendingDot?.()
      const defaultSlots = slots?.default?.() || []
      const pendingNode = pendingSlots || props.pending
      const pendingDotNode = pendingDotSlots || props.pendingDot || h(IxIcon, { name: 'loading' })
      let penddingItem: null | VNode = null

      if (pendingNode) {
        penddingItem = h(
          IxTimelineItem,
          { class: `${itemPrefixCls}-pending-dot` },
          { default: () => pendingNode, dot: () => pendingDotNode },
        )
      }

      const timelineItems = props.reverse ? [penddingItem, ...defaultSlots.reverse()] : [...defaultSlots, penddingItem]
      const nonNullItems = timelineItems.filter(item => !!item) as VNode[]
      const itemsLength = nonNullItems.length
      const items: VNode[] = []
      const itemPlacementArr: TimelinePlacement[] = []

      nonNullItems.forEach((item, index) => {
        const placement = getItemPlacement(item, index, props.placement)

        itemPlacementArr.push(placement)
        items.push(
          cloneVNode(item, {
            class: useItemClasses({
              hasPendingNode: !!pendingNode,
              placement,
              index,
              itemsLength,
              props: props,
            }),
          }),
        )
      })

      return h(
        'ul',
        {
          class: useClasses(itemPlacementArr, props),
        },
        items,
      )
    }
  },
})

const getItemPlacement = (node: VNode, index: number, placement: TimelinePlacement): TimelinePlacement => {
  const itemPlacement = node.props?.placement
  if (itemPlacement) {
    return itemPlacement
  }

  if (placement === 'alternate') {
    return index % 2 ? 'start' : 'end'
  }

  return placement
}

const getRealPlacement = (itemPlacementArr: TimelinePlacement[]): TimelinePlacement => {
  return itemPlacementArr.reduce((result: TimelinePlacement, nextPlacement, index) => {
    if (!index) {
      result = nextPlacement
    } else if (result !== nextPlacement) {
      result = 'alternate'
    }

    return result
  }, 'end')
}

const useClasses = (itemPlacementArr: TimelinePlacement[], props: TimelineProps): string => {
  const realPlacement = getRealPlacement(itemPlacementArr)
  const placementCls = `${timelinePrefixCls}-${realPlacement}`

  let cls = `${timelinePrefixCls} ${placementCls}`

  if (props.reverse) {
    cls += ` ${timelinePrefixCls}-reverse`
  }

  if (!props.both) {
    cls += ` ${timelinePrefixCls}-not-both`
  }

  return cls
}

const useItemClasses = ({
  hasPendingNode,
  placement,
  index,
  itemsLength,
  props,
}: {
  hasPendingNode: boolean
  placement: TimelinePlacement
  index: number
  itemsLength: number
  props: TimelineProps
}): string => {
  let cls = `${itemPrefixCls}-${placement}`

  if (hasPendingNode) {
    const isReversePending = props.reverse && index === 0
    const isPending = !props.reverse && index === itemsLength - 2

    if (isReversePending || isPending) {
      cls += ` ${itemPrefixCls}-pending`
    }
  }

  return cls
}
