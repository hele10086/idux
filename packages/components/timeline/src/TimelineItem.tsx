/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { CSSProperties } from 'vue'

import { computed, defineComponent } from 'vue'

import { hasSlot } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { isPresetColor, isStatusColor } from '@idux/components/utils'

import { timelineItemProps } from './types'

export default defineComponent({
  name: 'IxTimelineItem',
  props: timelineItemProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-timeline`)
    const prefixCls = mergedPrefixCls.value
    const isPresetOrStatus = computed(() => isPresetColor(props.color) || isStatusColor(props.color))

    const dotStyle = computed(() => {
      if (isPresetOrStatus.value) {
        return {}
      }
      return {
        color: props.color,
        'border-color': props.color,
      }
    })

    const dotClass = computed(() => {
      return {
        [`${prefixCls}-item-dot`]: true,
        [`${prefixCls}-item-dot-custom`]: hasSlot(slots, 'dot') || !!props.dot,
        [`${prefixCls}-item-dot-${props.color}`]: isPresetOrStatus.value,
      }
    })

    return () => {
      return (
        <li class={`${prefixCls}-item`}>
          <div class={`${prefixCls}-item-line`}></div>
          <div class={dotClass.value} style={dotStyle.value as CSSProperties}>
            {slots.dot ? slots.dot() : props.dot}
          </div>
          <div class={`${prefixCls}-item-content`}>
            <div class={`${prefixCls}-item-label`}>{props.label || slots.label?.()}</div>
            <div class={`${prefixCls}-item-desc`}>{slots.default?.()}</div>
          </div>
        </li>
      )
    }
  },
})
