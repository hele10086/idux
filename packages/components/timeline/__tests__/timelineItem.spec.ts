import { MountingOptions, mount } from '@vue/test-utils'

import { renderWork } from '@tests'

import IxTimelineItem from '../src/TimelineItem'
import { TimelineItemProps } from '../src/types'

describe('TimelineItem', () => {
  const timelineItemMount = (options?: MountingOptions<Partial<TimelineItemProps>>) =>
    mount(IxTimelineItem, { ...options })
  renderWork(IxTimelineItem)

  test('color work', async () => {
    const wrapper = timelineItemMount({
      props: {
        color: 'red',
      },
    })
    expect(wrapper.find('.ix-timeline-item-dot').classes()).toContain('ix-timeline-item-dot-red')
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({
      color: 'black',
    })

    expect(wrapper.find('.ix-timeline-item-dot-black').exists()).toBe(false)
    expect(wrapper.find('.ix-timeline-item-dot').html()).toContain('style="color: black; border-color: black;"')
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('dot work', async () => {
    const dotText = 'dotText'
    const wrapper = timelineItemMount()
    expect(wrapper.find('.ix-timeline-item-dot-custom').exists()).toBe(false)
    expect(wrapper.find('.ix-timeline-item-dot').text()).toBe('')
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({
      dot: dotText,
    })
    expect(wrapper.find('.ix-timeline-item-dot-custom').exists()).toBe(true)
    expect(wrapper.find('.ix-timeline-item-dot').classes()).toContain('ix-timeline-item-dot-custom')
    expect(wrapper.find('.ix-timeline-item-dot').text()).toBe(dotText)
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('dot slot work', async () => {
    const dotSlotText = 'dotSlotText'
    const dotText = 'dotText'
    const wrapper = timelineItemMount({
      slots: {
        dot: dotSlotText,
      },
    })
    expect(wrapper.find('.ix-timeline-item-dot-custom').exists()).toBe(true)
    expect(wrapper.find('.ix-timeline-item-dot').classes()).toContain('ix-timeline-item-dot-custom')
    expect(wrapper.find('.ix-timeline-item-dot').text()).toBe(dotSlotText)
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({
      dot: dotText,
    })
    expect(wrapper.find('.ix-timeline-item-dot').text()).toBe(dotSlotText)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
