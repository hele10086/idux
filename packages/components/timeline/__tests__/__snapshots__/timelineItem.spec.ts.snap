// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineItem color work 1`] = `
"<li class=\\"ix-timeline-item\\">
  <div class=\\"ix-timeline-item-line\\"></div>
  <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-red\\">
    <!---->
  </div>
  <div class=\\"ix-timeline-item-content\\">
    <div class=\\"ix-timeline-item-label\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-desc\\">
      <!---->
    </div>
  </div>
</li>"
`;

exports[`TimelineItem color work 2`] = `
"<li class=\\"ix-timeline-item\\">
  <div class=\\"ix-timeline-item-line\\"></div>
  <div class=\\"ix-timeline-item-dot\\" style=\\"color: black; border-color: black;\\">
    <!---->
  </div>
  <div class=\\"ix-timeline-item-content\\">
    <div class=\\"ix-timeline-item-label\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-desc\\">
      <!---->
    </div>
  </div>
</li>"
`;

exports[`TimelineItem dot slot work 1`] = `
"<li class=\\"ix-timeline-item\\">
  <div class=\\"ix-timeline-item-line\\"></div>
  <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-custom ix-timeline-item-dot-primary\\">dotSlotText</div>
  <div class=\\"ix-timeline-item-content\\">
    <div class=\\"ix-timeline-item-label\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-desc\\">
      <!---->
    </div>
  </div>
</li>"
`;

exports[`TimelineItem dot slot work 2`] = `
"<li class=\\"ix-timeline-item\\">
  <div class=\\"ix-timeline-item-line\\"></div>
  <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-custom ix-timeline-item-dot-primary\\">dotSlotText</div>
  <div class=\\"ix-timeline-item-content\\">
    <div class=\\"ix-timeline-item-label\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-desc\\">
      <!---->
    </div>
  </div>
</li>"
`;

exports[`TimelineItem dot work 1`] = `
"<li class=\\"ix-timeline-item\\">
  <div class=\\"ix-timeline-item-line\\"></div>
  <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
    <!---->
  </div>
  <div class=\\"ix-timeline-item-content\\">
    <div class=\\"ix-timeline-item-label\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-desc\\">
      <!---->
    </div>
  </div>
</li>"
`;

exports[`TimelineItem dot work 2`] = `
"<li class=\\"ix-timeline-item\\">
  <div class=\\"ix-timeline-item-line\\"></div>
  <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-custom ix-timeline-item-dot-primary\\">dotText</div>
  <div class=\\"ix-timeline-item-content\\">
    <div class=\\"ix-timeline-item-label\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-desc\\">
      <!---->
    </div>
  </div>
</li>"
`;

exports[`TimelineItem render work 1`] = `
"<li class=\\"ix-timeline-item\\">
  <div class=\\"ix-timeline-item-line\\"></div>
  <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
    <!---->
  </div>
  <div class=\\"ix-timeline-item-content\\">
    <div class=\\"ix-timeline-item-label\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-desc\\">
      <!---->
    </div>
  </div>
</li>"
`;
